### PR TITLE
Update to Babel 6 (needs further work!)

### PR DESCRIPTION
**Pull request by bjrn**
So, I am aware that this PR is doing _way_ too much at the moment ... and the component documentation is broken, likely due to that `react-component-metadata` needs work in order to be compatible with Babel 6.

I worked around the issue by wrapping the metadata call with a try/catch in `docs/generate-metadata.js`
This results in that `window.PROP_DATA` in `docs/src/Root.js` is empty, thus the component page is disfunctional. 

Besides the update to use Babel 6, there's fixes for updated react-router, linting of a few files to fix the build etc. 

I would really appreciate if someone could give some pointers on how to move this along! Am I right in the assumption that `react-component-metadata` needs to be patched as well?

### CODE REVIEW

**Comment by jquense** [comment]
thanks for working on this. you are right about react-component-metadata. Luckily I updated it a few days ago to not need a Babel peer  any more so it will work with Babel 6 now, you do need to update it to the latest version tho.

File: docs/src/ReactPlayground.js [hunk 1]
```diff
@@ -61,12 +61,22 @@ const Tooltip = require('../../src/Tooltip');
...
+  ]
+};
+
+// import CodeExample from './CodeExample';
+const CodeExample = require('./CodeExample');
 
 // This is only used for the ReactPlayground component, not for any of the
 // examples, so it's fine to import like this.
-import SafeAnchor from '../../src/SafeAnchor';
+// import SafeAnchor from '../../src/SafeAnchor';
```
Comment (jquense) [inline docs/src/ReactPlayground.js]: why change the import style here?

File: package.json [hunk 2]
```diff
@@ -111,7 +119,9 @@
     "yargs": "^4.3.2"
   },
   "dependencies": {
-    "babel-runtime": "^5.8.38",
+    "babel-plugin-transform-decorators-legacy": "^1.3.4",
+    "babel-runtime": "^6.6.1",
+    "babel-standalone": "^6.7.4",
```
Comment (jquense) [inline package.json]: both the plugin and the babel-standalone should go in devDependencies

File: .babelrc [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
+  "presets": [
+    "react",
+    "es2015",
+    "stage-1"
+  ],
+
+  "plugins": [
+    "transform-runtime",
+    "add-module-exports",
+    "transform-decorators-legacy",
```
Comment (jquense) [inline .babelrc]: I think I'd prefer to just refactor the two places we use decorators to not use them, lowers complexity and makes it more consistent with the rest of the codebase anyway

**Comment by bjrn** [comment]
ok, I'm using version 3.0.0 from npm in the branch but it's still failing...

File: package.json [hunk 4]
```diff
@@ -93,12 +101,12 @@
     "output-file-sync": "^1.1.1",
     "phantomjs-prebuilt": "^2.1.7",
     "portfinder": "^1.0.3",
-    "react": "^15.0.0",
-    "react-component-metadata": "^2.1.1",
-    "react-dom": "^15.0.0",
+    "react": "^15.0.1",
+    "react-component-metadata": "^3.0.0",
```
Comment (jquense) [inline package.json]: It looks like you have the latest major version here, that should work just fine with Babel6

File: src/Modal.js [hunk 5]
```diff
@@ -142,11 +142,11 @@ const Modal = React.createClass({
 
   render() {
     let {
-        className
-      , children
-      , dialogClassName
-      , animation
-      , ...props } = this.props;
+      className,
```
Comment (jquense) [inline src/Modal.js]: where is this refactor coming from?

File: test/utils/bootstrapUtilsSpec.js [hunk 6]
```diff
@@ -191,7 +192,8 @@ describe('bootstrapUtils', ()=> {
     });
 
     it('should work with createClass', ()=> {
-      shouldWarn('expected one of ["smallish","micro","planet"]');
+      // this triggers an error: the warning list is empty
+      // shouldWarn('expected one of ["smallish","micro","planet"]');
```
Comment (jquense) [inline test/utils/bootstrapUtilsSpec.js]: I think this is unrelated to Babel update, I would make sure you've rebased against the latest master

File: src/Modal.js (hunk 5 above)
Comment (bjrn) [inline src/Modal.js]: I got linting errors from that one (and the lint errors stopped the build) so I changed it to comply. Like I said in the PR Comment, this PR is doing way too much ;)

File: test/utils/bootstrapUtilsSpec.js (hunk 6 above)
Comment (bjrn) [inline test/utils/bootstrapUtilsSpec.js]: Yep, it is most likely unrelated to Babel, probably due to that I ran an `npm-check-updates` on the branch...

**Comment by jquense** [comment]
I'd make sure you are using the absolute latest, and I might blow away node_modules and reinstall to make sure nothing weird is happening there

File: .babelrc [hunk 7]
```diff
@@ -1,6 +1,15 @@
...
+    "dev-expression",
+    "transform-es3-modules-literals"
```
Comment (taion) [inline .babelrc]: Consider using the airbnb preset, that includes all the es3 stuff

File: package.json (hunk 4 above)
Comment (bjrn) [inline package.json]: @jquense seems like `react-component-metadata 3.0.0` is incompatible with the react-bootstrap components when they are transpiled with Babel 6. Any chance you could give some pointers on how to make an isolated test case for this?

Comment (jquense) [inline package.json]: can you tell me what the error is?

Comment (bjrn) [inline package.json]: @jquense Here's the error log:
```
Error: Command failed: npm run docs-build -- 
TypeError: Converting circular structure to JSON
    at Object.stringify (native)
    at Object.stringify (/Users/bjorn/dev/react-bootstrap/node_modules/core-js/modules/es6.symbol.js:138:21)
    at stringify (/Users/bjorn/dev/react-bootstrap/node_modules/core-js/library/fn/json/stringify.js:4:26)
    at [object Object].render (Root.js:36:30)
    at [object Object].ReactCompositeComponentMixin._renderValidatedComponentWithoutOwnerOrContext (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactCompositeComponent.js:679:34)
    at [object Object].ReactCompositeComponentMixin._renderValidatedComponent (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactCompositeComponent.js:699:32)
    at [object Object].wrapper [as _renderValidatedComponent] (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactPerf.js:66:21)
    at [object Object].ReactCompositeComponentMixin.performInitialMount (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactCompositeComponent.js:284:30)
    at [object Object].ReactCompositeComponentMixin.mountComponent (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactCompositeComponent.js:237:21)
    at [object Object].wrapper [as mountComponent] (/Users/bjorn/dev/react-bootstrap/node_modules/react/lib/ReactPerf.js:66:21)
```

Looks like it gets the component metadata all wrong...

Comment (jquense) [inline package.json]: hmm that doesn't seem to explicity be about metadata...

I have a branch with half of babel6 support in it let me try that and see if I get something similar

Comment (bjrn) [inline package.json]: ok, sounds great!

Comment (bjrn) [inline package.json]: Hi @jquense, did you have time to try on your own local branch?

**Comment by ruiaraujo** [comment]
@bjrn @taion Any progress on this? Could I try doing it myself?

**Comment by bjrn** [comment]
Hi @ruiaraujo, not much progress I'm afraid — I took another stab at this, where I started from the beginning again -  but never succeeded to get it to a point where it all worked. That branch is probably a better starting point if you want to bring this further: https://github.com/bjrn/react-bootstrap/tree/babel6

I got it to a point where the library compiled and all, but the live-editable styleguide threw all sorts of errors. Feel free to have a go!

**Comment by jknight12882** [comment]
+1 on this. having a `babel-runtime` for `5.x.x` from this project is doubling the amount of dependencies for my `Babel 6.x.x` project.

**Comment by LeoAJ** [comment]
Got same error

screen shot 2016-07-08 at 5 34 58 pm

If I remove building doc it will work fine, it looks like `building doc` is the only blocker for this one.

**Comment by taion** [comment]
-> #2108
